resources/rules.k.tmpl

```diff
@@ -375,19 +375,19 @@ rule ACCTCODE in SetItem( 1 )
     rule  A -Int ( (B /Int 64)  +Int C) => (0 -Int ( (B /Int 64) -Int (A -Int C)))
       requires (notBool #isVariable(A))
       andBool (notBool #isConcrete(A))
-      andBool (A =/=K (D /Int 64))
+      andBool (A =/=K (B /Int 64))
```

Comment:
These look OK to me.

Reply:
yeah, what was before made no sense. This could be related to the build failure though

Reply:
sry but `D` or something similar is the right choice here.
The rule is there to create and clean an arithmetic normal form to be able to simplify the gas formula:
whenever we have a an arithmetic expression `(A + B) - (C + D)` those normal form rules looks to make a right associative list out of it. e.g. `(((A + B) - C) - D)`  those lists can be arbitrary in length.
We additionaly have simplification rules of the form `A + B - B => A` but a bit more complicated.
Here we can't find a B expression occurring positively and negatively at arbitrary height in the list with the current implementation, we always have a limited lookahead (1-2) so we need to _sort_ the list to move expressions of the same form to either position and 'hope' that it works our lookahead is sufficient to simplify them in practice. e.g. all variables to the left, all concrete values to the right, all expressions of the form `X / 64` to the right of all variables but left of everything else. Now what happenes here with the new rule if `A == D / 64` with `D =/= B`?!
then the new rule will apply and reshuffle them. but note that the new reshuffled form has again the same form and will lead to an infinite reapplication of the same rule.
The previous implementation statet that D can be arbitrary, we don't care about it but A is not of the form `SOMETHING / 64`
I'm not sure how to implement it otherwise

Reply:
@mhhf I think it would be great if this were documented in the file

Reply:
Thanks a ton for the context, @mhhf! Sounds the like the correct thing is then to replace `D` with `?D` to conform to the new `K` requirements. I've made that change locally and assuming a few proof spot-checks succeed, I'll update the PR. Agree with @asymmetric that it would be great to have this documented somewhere.

Reply:
Update pushed, but I do have a further question here:

These rules have
```
      requires (notBool #isVariable(A))
      andBool (notBool #isConcrete(A))
```

i.e. `A` must be neither concrete nor symbolic. I asked Everett about this earlier today, and according to him one of `isVariable` and `isConcrete` should return true for any expression. Based on that, these rules will never be applied. Can someone comment on this (@mhhf or @MrChico, maybe?)?

Reply:
i'm not sure if they changed it but for example `A ==K X /Int 64` is neither a variable or a concrete value. `X` is a variable and `64` concrete but the whole expression is a `KApply`/ complex symbolic expression which is intended to be catched by those constraints.

Reply:
I would think that every expression is either `#isSymbolic` or `#isConcrete`, but `#isVariable` is a proper subset of `#isSymbolic`.